### PR TITLE
(MODULES-9304) Add Puppet Strings docs

### PR DESCRIPTION
**Pull request by eimlav**

### CODE REVIEW

**Comment by eimlav** [comment]
https://github.com/puppetlabs/puppetlabs-acl/blob/76fdb70f995de47f5e99f7a6ef580abfb24ef5a4/lib/puppet/provider/acl/windows/base.rb#L17-L21

The reason for this change is that by putting everything inside the if block, YARD is unable to parse the comments of any of the methods. I tried a number of different things to remove this if block however every time it would result in errors. By applying it only on the requires, the tests pass fine however I'm not sure if this is the best solution. Feel free to give some other ideas of better alternatives.

File: lib/puppet/provider/acl/windows/base.rb [hunk 1]
```diff
@@ -1,477 +1,588 @@
...
 
-            if (ace.mask & FILE_GENERIC_READ) == FILE_GENERIC_READ
-              rights << :read
-              mask_specific_remainder &= ~FILE_GENERIC_READ
-            end
-            if (ace.mask & GENERIC_READ) == GENERIC_READ
-              rights << :read
-              mask_specific_remainder &= ~GENERIC_READ
-            end
+      # Retrieves the TODO
```
Comment (RandomNoun7) [inline lib/puppet/provider/acl/windows/base.rb]: I don't understand this line.

Comment (eimlav) [reply lib/puppet/provider/acl/windows/base.rb]: Updated.

File: lib/puppet/type/acl/ace.rb [hunk 2]
```diff
@@ -124,12 +164,18 @@ def ensure_rights_values_compatible
...
     def ensure_mask_when_mask_specific
       if @rights.include?(:mask_specific) && (@mask.nil? || @mask.empty?) # rubocop:disable Style/GuardClause  Changing this to a guard clause makes the line long and unreadable
         raise ArgumentError, "If you specify rights => ['mask_specific'], you must also include mask => 'value'. Reference: #{inspect}"
       end
     end
 
+    # Checks that a supplied array of values are unique
+    #
+    # @param [Array] values
+    # @return [Boolean] If values are unique or not
```
Comment (RandomNoun7) [inline lib/puppet/type/acl/ace.rb]: I don't think the return here is going to be a bool. It's either going to return a new array with only unique values, or if it's not an array return to supplied value.

File: lib/puppet/type/acl/ace.rb [hunk 3]
```diff
@@ -308,14 +390,23 @@ def keys
...
+    # Returns values of instance fields
+    #
+    # @return [Array] Values of fields from hash
     def values
       to_hash.values
     end
 
+    # Returns value of specified key in instance fields hash
+    #
+    # @return [Object] Value of field
```
Comment (RandomNoun7) [inline lib/puppet/type/acl/ace.rb]: Since to_hash is using string keys I think the input param here also has to be a string?

**Comment by glennsarti** [comment]
>  By applying it only on the requires, the tests pass fine however I'm not sure if this is the best solution. Feel free to give some other ideas of better alternatives.

Strictly speaking you should probably add a guard clause to every single public method e.g.

```
def do_something()
  raise "Only Supported on Windows" unless Puppet::Util::Platform.windows?
  ...
end
```

However...this whole if thing _could_ be a hangover from the Puppet 3 days and the tests may be testing the wrong thing?

**Comment by eimlav** [comment]
@glennsarti I believe the errors are to with Windows platform specific gem dependencies that do not work on nix based systems (i.e. the gems being required in the file). The acceptance test suite also contains similar if clauses for requiring gems from `puppet/util/windows/...`. For now the proposed solution appears to work in adhoc however finding a cleaner solution is something that should be investigated in the future.

**Comment by RandomNoun7** [comment]
Manually tested the module with a Puppet master deliberately trying to trigger an explosion. Even with no `if` clause it all, the module did not cause an error during catalog compilation, and it functioned as expected once the code ran on the Windows node.

**Pull request merged**